README
This is the fifth public version of Cockatrice (update six?)

This version is compiled with GCC 8.1.0 with the -Os flag, as distributed as:
i686-8.1.0-win32-sjlj-rt_v6-rev0

 **** THERE IS A BUG IN PARSING PREF FILES, PUT AN EMPTY LINE or a HASH as the 
last entry, and everything will be fine!!!!!!!!!!!!!! ****

WHATS NEW?  Well the bigges thing is a SDL sound change from rakslice fixing
audio playback of when MacOS would pick strange formats, and not trying to match
it, but forcing SDL to stick to one format as needed.  Real Audio, Retrun to Zork,
and many more games now work (Including flash player, KHAAAN!!!)

I also added the ability to force the clock behind allowing timebombed installers
to live like it's 1998

Ive tried to iron out a few prefs bugs, so some of the mysterious 
crashing should be over.  Ive also integrated DrMinGW in this Win32
build to dump out some debug logs if there is any issues.  I've also change
the GCC to build -Ofast as the new recommended level for building neutral
fast applications, and include all the debug information.
Because it's compiled in, it won't upset your normal JIT debugger (assuming
you even have one).

In any event, if you are having issues, give me the CPU type, OS, and 
the rpt file, and I can at least have a guess as to why it's crashing.

Enclosed is two executables, the SDL dll that I'm using 
(although there is nothing special about it), and all the DLLs associated
with Dr MinGW.

08/22/2013  05:03 AM           124,520 symsrv.dll
02/20/2014  10:13 AM         1,256,112 dbghelp.dll
11/12/2015  06:49 AM           216,064 mgwhelp.dll
11/12/2015  06:49 AM            23,040 exchndl.dll
12/19/2018  01:49 PM           339,968 SDL.dll
03/24/2010  11:32 AM           224,768 qemu-img.exe
02/10/2016  01:01 AM            35,790 ethlist.exe
10/28/2020  01:46 PM         3,914,844 CockatriceIII.exe

To create a new disk image, I've enclosed the qemu-img program.
Just run it something like this:

qemu-img.exe create -f raw scsi0.disk 2000M

and it'll create a 2,000 MB file suitable for Cockatrice III

To further aid people in getting started I threw in a minimal
System 7 & System 8 install in the archive directory.

10/02/2016  01:09 AM           300,061 blank_2gb.7z
09/02/2016  11:53 PM         1,896,688 min7.7z
09/02/2016  11:59 PM         8,868,401 min8.7z

These will extract out to scsi0.disk, so don't overwrite that unless
you want to.

The previous release of Cockatrice, greatly improved
networking speed and stability.  I can now copy several GB worth of 
data over AppleTalk, while before this was dreadfully slow and unstable.
I've also started to expand the SCSI code to better support the buffer
demands of Basilisk II, so you can now format disks 2GB without it
crashing.  I have also started to add primative CD-ROM support, in addition
to the SCSI disk files.  The SCSI device #6 is hard coded to be a CD-ROM.
There is no 'hot' insertation method, and ignore/lock are ignored for now.
Simply use the following statement in your config file:

scsi6 <diskfilename>

But at this time, ISO-9660 filesystems will not mount.  Only *SOME*
HFS volumes will.  I have also made available on sourceforge a 'util' CD
that contains among other things the disk tool from A/UX that can format
any disk.  You can find it at:

https://sourceforge.net/projects/cockatrice/files/Utils/

I highly recommend using it if you are starting for the first time.  To
force your virtual MAC to boot from the CD use the statements:

disk6 CDUtils.iso
bootdriver -33
bootdisk 6

And this will ensure that the CD-ROM gets boot priority.  I've found that
the Macintosh Quadra 800 ROM is the best for running System 7 and 8.

My BasiliskII_Prefs is as follows:

-------------------------------------------------------------------
#screen win/1152/870
screen win/1000/600
#ether slirp
ether {FFFFFFFF-FFFF-7777-FFFF-FFFFFFFFFFFF}
rom Quadra800.rom
bootdrive 0
bootdriver -33
#ramsize 67108864
ramsize 268435456
frameskip 2
modelid 29
cpu 4
fpu true
nocdrom false
nosound false
nogui false
idlewait false
scsi0 OS7.disk
scsi6 CDUtils.iso
disk xfer.dsk
#disk rtz.disk
yearoffset 7
#
-------------------------------------------------------------------


This way my SCSI disk boots instead of trying to boot from the 
transfer.hfv volume, which is the default behavour.

is the syntax to add a disk into position 0 of the SCSI bus.

Another 'feature' is that for 68040 emulation you can now disable the
numerical coprocessor.  You may find this desierable as it is quite buggy
and things like the scrollbars will function in OS 8.

= Networking
As always, pcap is present, This will allow you to  use native networking.  
Although this is more geared to wired networking.  For wireless, use SLiRP.  
There is no tun/tap networking in this version.

To setup the Windows networking, run the program ethlist to determin the 
network adapters available in your computer.


C:\BasiliskII\mingw>ethlist.exe
Network devices:
  Number       NAME                                     (Description)
  0  {FFFFFFFF-FFFF-7777-FFFF-FFFFFFFFFFFF} (Local Area Connection)

Press Enter to continue...


In this example there is a single interface.  But due to the nature of 
Windows, the index number could easily change, but the GUID is what pcap 
is expecting and for now that is what you have to specifiy in the 
BasiliskII_Prefs file.

In this example you would add the line:

ether {FFFFFFFF-FFFF-7777-FFFF-FFFFFFFFFFFF}

And then you shold be good to go.

When you start Cockatrice, you should then see something similar to this:

C:\BasiliskII\mingw>CockatriceIII.exe
Basilisk II V0.5 by Christian Bauer et al.
Reading ROM file...
Setting up for a 68040, With FPU and 32bit addressing
SCSI Unit 0:scsi0.disk geometry: 50 sectors, 2621440 cylinders, 16 heads
EtherInit with [PCAP]
Pcap version [WinPcap version 4.1.3 (packet.dll version 10, 2, 0, 5002), based on libpcap version 1.0 branch 1_0_rel0b (20091008)]
Ethernet address fe fd 00 00 6e 60
Building packet filter.........!
SDL_Audio inited 16bit, 44100Hz, 2 channels
SDL_Video 600x1000 8bit deep/size 1
Patching a 32-bit clean ROM (version $067c or higher)

On Windows 10 with the newer pcap it will constantly print the following message:

'FATAL Bad Memory Block.'

However as far as I can tell this is 'cosmetic' aka non fatal message from WinPcap.


Linux and OS X users just need to specify the name of their 
ethernet interfaces.  WIFI ADAPTERS WILL NOT WORK!!!  Instead use SLiRP.

You may also need to run this as the root user.  In a future release I'll
try a stab at the user mode support, although I think that is a Windows
only thing.

= SLiRP
To use the SLiRP (ie user mode NAT) you simply use the statement

ether slirp

And if your MacOS uses DHCP it'll figure out the configuration 
on it's own.  Otherwise use the folowing address:

IP			10.0.2.15
NetMask			255.255.255.0
Gateway			10.0.2.2
DNS			10.0.2.3

YOU CANNOT PING ANYTHING ON YOUR NETWORK OR THE INTERNET 
with SLiRP mode.  This is normal, and expected.  If you must ping something 
to ensure the stack is working, ping 10.0.2.2.  It is the only thing that can ping.

I've done my best to make the SLiRP as stable as possible, but it's possible 
that it may fail frm time to time. 

But I've downloaded 150MB files at 450KB/sec without issue.  

Plenty of things are missing from this version, as it is derived from 
BasiliskII version 0.8.  I am striving to make it 100% SDL reliant for 
system calls, but for now the os x/Linux and MinGW are more tied to UNIX, 
while the Visual C++ version is more agnostic.  But for now I'm only 
happy with the GCC/Clang built versions. But the Visual C++ build is making
great progress

I've mostly tested with os 8.0 for internet stuff, and 7.11 for using 
AppleTalk to communicate with Shoebill, and Windows NT AppleTalk servers.


For System 7.1.1 without AppleTalk or OpenTransport I just set the modelid
to 5, and restrict the CPU to 3

Various other ROMs either have booting issues from SCSI, or other errors that
BasiliskII's ROM patches cannot compensate for.

= Y2K and going back in time.
Installers like RealAudio only work in the late 90's they will refuse to work.
I added a config line of:
yearoffset 0

Change it to any number and it'll remove that many billion ticks away from
the offset allowing you to travel back in time.
As of 2020 a 7 sends me back to 1998.  Maybe if there is enough need I can
refine it to setting the year/month/date...

-------
C:\Cockatrice\temp\Cockatrice-v0.0-5>gcc -v
Using built-in specs.
COLLECT_GCC=gcc
COLLECT_LTO_WRAPPER=C:/Program\ Files\ (x86)/mingw-w64/i686-8.1.0-win32-sjlj-rt_v6-rev0/mingw32/bin/../libexec/gcc/i686-w64-mingw32/8.1.0/lto-wrapper.exe
Target: i686-w64-mingw32
Configured with: ../../../src/gcc-8.1.0/configure --host=i686-w64-mingw32 --build=i686-w64-mingw32 --target=i686-w64-mingw32 --prefix=/mingw32 --with-sysroot=/c/mingw810/i686-810-win32-sjlj-rt_v6-rev0/mingw32 --enable-shared --enable-static --enable-targets=all --enable-multilib --enable-languages=c,c++,fortran,lto --enable-libstdcxx-time=yes --enable-threads=win32 --enable-libgomp --enable-libatomic --enable-lto --enable-graphite --enable-checking=release --enable-fully-dynamic-string --enable-version-specific-runtime-libs --enable-sjlj-exceptions --disable-libstdcxx-pch --disable-libstdcxx-debug --enable-bootstrap --disable-rpath --disable-win32-registry --disable-nls --disable-werror --disable-symvers --with-gnu-as --with-gnu-ld --with-arch-32=i686 --with-arch-64=nocona --with-tune-32=generic --with-tune-64=core2 --with-libiconv --with-system-zlib --with-gmp=/c/mingw810/prerequisites/i686-w64-mingw32-static --with-mpfr=/c/mingw810/prerequisites/i686-w64-mingw32-static --with-mpc=/c/mingw810/prerequisites/i686-w64-mingw32-static --with-isl=/c/mingw810/prerequisites/i686-w64-mingw32-static --with-pkgversion='i686-win32-sjlj-rev0, Built by MinGW-W64 project' --with-bugurl=https://sourceforge.net/projects/mingw-w64 CFLAGS='-O2 -pipe -fno-ident -I/c/mingw810/i686-810-win32-sjlj-rt_v6-rev0/mingw32/opt/include -I/c/mingw810/prerequisites/i686-zlib-static/include -I/c/mingw810/prerequisites/i686-w64-mingw32-static/include' CXXFLAGS='-O2 -pipe -fno-ident -I/c/mingw810/i686-810-win32-sjlj-rt_v6-rev0/mingw32/opt/include -I/c/mingw810/prerequisites/i686-zlib-static/include -I/c/mingw810/prerequisites/i686-w64-mingw32-static/include' CPPFLAGS=' -I/c/mingw810/i686-810-win32-sjlj-rt_v6-rev0/mingw32/opt/include -I/c/mingw810/prerequisites/i686-zlib-static/include -I/c/mingw810/prerequisites/i686-w64-mingw32-static/include' LDFLAGS='-pipe -fno-ident -L/c/mingw810/i686-810-win32-sjlj-rt_v6-rev0/mingw32/opt/lib -L/c/mingw810/prerequisites/i686-zlib-static/lib -L/c/mingw810/prerequisites/i686-w64-mingw32-static/lib -Wl,--large-address-aware'
Thread model: win32
gcc version 8.1.0 (i686-win32-sjlj-rev0, Built by MinGW-W64 project)

ChangeLog
v0.5a 28.Oct.2020
	- Added a SDL audio fix from rakslice
	- Added the ability to subtract from the clock to go to 1998
	- Rebuilt using GCC 8.1.0 on MinGW.  Changed flags to -Os

v0.4g 15.May.2016
	- Fixed some stupid mappings in the keyboard
	- Just forced the sound to 16bit 44100, and tested with Quicktime & 
	  Simcity 2000.
	- added in some branding elemnets, a Windows icon.
	- changed the version to reflect Cockatrice III
	- changed the perfs file, and updated default prefrences to my
	  test setup (Quadra 800)
	- Building with GCC 6.1.0 on MinGW.  Changed flags to -Ofast
	  as it was the fastest generic flags I could see

v0.0 (release 0.0-3) - 17.Sep.2015
	- Corrected stupid SLiRP bug, introduced by me. :(

v0.0 (release 0.0-2) - 14.Oct.2014
	- More mutexes and fixes on slirp.
	- some improvements in blocking and timings with PCAP
	- Imported SCSI driver from previous.  Now SCSI disk files work

V0.0 (release 0.0-1) - 09.Oct.2014
	- pruned the serial interface
	- told the disk driver to scan for partitions
	- more fixes for slirp, regarding a mutex around all functions
	  in sdl_pcap.cpp

V0.0 (release 0.0-0) - 07.Oct.2014
	- first release showing off preliminary SDL video,sound and input
	- Also includes libpcap for raw networking (AppleTalk!)
	- I've also encorporated SLiRP from later Basilisk II
	- Name Change and new life... welcome to Cockatrice III

V0.8 (release 0.8-1) - 30.Jan.2000
	- extfs: extended size of finf helper files to 32 bytes to eventually
	  allow storing the complete FInfo/FXInfo in it

V0.8 (snapshot) - 22.Dec.1999
	- imported UAE CPU 0.8.10 changes
	- new utility functions Mac_memset, Mac2Host_memcpy, Host2Mac_memcpu
	  and Mac2Mac_memcpy
	- extfs.cpp: fixed bug in fs_rename() and fs_cat_move() (auxiliary
	  IOParam block was not in Mac address space and FSItem parent IDs
	  didn't get swapped) [Lauri Pesonen]
	- extfs.cpp: several fixes [Lauri Pesonen]
	- some provisions for using UAE CPU compiler (doesn't work yet)
	- video_x.cpp: XFree86 VidMode extension is used to switch modes
	  in DGA mode

V0.8 (snapshot) - 2.Nov.1999
	- fixed many UAE FPU bugs [Lauri Pesonen]
	- added replacement for BlockMove() trap
	- removed Windows sources from the source archive; a version of
	  these that actually compiles and works can be downloaded from
	  Lauri Pesonen's site
	- removed the MemoryDispatch() replacement routine; the routine
	  in the ROM is now always used; this fixes the MacOS 8 Finder
	  copying bug and also most "unimplemented trap" errors on MacOS
	  bootup; yeah :-)
	- a warning is issued if the ROM seems to require an FPU but FPU
	  emulation is turned off
	- medium removal is allowed in CDROMExit()
	- added (incomplete) emulation of 68040 instructions (CINV, CPUSH,
	  MOVE16 (Ax)+,(Ay)+, and FPU stack frames), enough to boot MacOS
	- UAE CPU: opcode routines no longer return the cycle count
	- extfs.cpp: fixed bug with fsResolveWDCB in fs_get_wd_info()
	- Unix: added support for ESD audio output; merged with OSS audio
	  and put in a new "audio_oss_esd.cpp" file which is also used under
	  FreeBSD 3.x
	- Unix: added mkinstalldirs to "make install" target
	- Unix: cleaned up the configure script
	- Unix: ROM breakpoint can be specified with "-break" command line
	  argument
	- Unix: "-rominfo" command line argument to print information
	  about ROM version and tables
	- Unix: ExtFS supports resource forks and Finder info (these are
	  kept in hidden ".finf" and ".rsrc" directories)
	- Unix/audio_oss_esd.cpp: AudioStatus is re-set after determining
	  audio device capabilities (actual sample rate is also determined)
	  [Alexander R. Pruss]
	- Unix/audio_oss_esd.cpp: "silence" in 8-bit mode used wrong fill
	  value (0 instead of 0x80) [Alexander R. Pruss]
	- Unix/video_x.cpp: added mouse wheel support [Alexander R. Pruss]
	- Unix/sys_unix.cpp: device size is read correctly under Linux
	- AmigaOS: removed support for SAS/C
	- AmigaOS/sys_amiga.cpp: supports 64-bit device I/O (NSD)
	- AmigaOS/sys_amiga.cpp: device block size is respected on writes

V0.8 (snapshot) - 21.Oct.1999
	- sony.cpp/disk.cpp/cdrom.cpp: disk insertions are now checked for
	  by an interrupt routine
	- localizable strings are now split into a common and a platform-
	  specific set
	- added external file system to access host OS files from the MacOS;
	  root directory is specified by the "extfs" prefs item
	- moved most initialization/deinitialization code to InitAll() and
	  ExitAll() in main.cpp
	- added patches for NetBSD [Bernd Sieker]
	- corrected TimerDateTime() in timer_unix.cpp and timer_beos.cpp
	  [Toshimitsu Tanaka]
	- AmigaOS: fixed crash when gtlayout.library was not present
	- AmigaOS: now compiles with GCC (Geek Gadgets)
	- AmigaOS/clip_amiga.cpp: fixed small bug in CR->LF translation
	  [Giacomo Magnini]
	- Unix: compilation of cpuemu.cpp is now split in 8 parts
	- Unix: volume list in GTK prefs editor is reorderable
	- Unix: added fbdev DGA support and SPARC assembly optimizations to
	  UAE core [Gwenole Beauchesne]

V0.7 (release 0.7-2) - 6.Oct.1999
	- Added BasiliskII.spec for making RPMs [with assistance from
	  Roman Niewiarowski]
	- Unix: "basilisk_ii_keycodes" renamed to "keycodes", default
	  location of keycodes file is now /usr/local/share/BasiliskII/keycodes
	- Unix: improved the install target, added uninstall target
	- Unix/Windows: fixed bugs in protocol removal [Lauri Pesonen]
	- Unix/Linux/NetDriver: removed MAKEDEV, functionality now included
	  in Makefile ("make dev")
	- Unix/sys_unix.cpp: Sys_open() checks whether an HFS volume is currently
	  mounted under Linux and tries to unmount it
	- Unix/sys_unix.cpp: implemented SysAddDiskPrefs() for Linux (scans
	  /etc/fstab to find HFS volumes)
	- Unix/sys_unix.cpp: implemented Sys(Prevent/Allow)Removal() [Ryan Cumming]
	- Unix/main_unix.cpp: added call to gtk_set_locale() before gtk_init()
	  [Toshimitsu Tanaka]
	- Unix/video_x.cpp: "dga" screen preferences now takes width and

BasiliskII/prefs.cpp
/*
 *  prefs.cpp - Preferences handling
 *
 *  Basilisk II (C) 1997-1999 Christian Bauer
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

#include <string.h>
#include <stdlib.h>
#include <stdio.h>
#include <ctype.h>

#include "sysdeps.h"
#include "sys.h"
#include "prefs.h"


// Common preferences items (those which exist on all platforms)
// Except for "disk", "floppy", "cdrom", "scsiX", "screen", "rom" and "ether",
// these are guaranteed to be in the prefs; "disk", "floppy" and "cdrom" can
// occur multiple times
prefs_desc common_prefs_items[] = {
	{"disk", TYPE_STRING, true},		// Device/file names of Mac volumes (disk.cpp)
	{"floppy", TYPE_STRING, true},		// Device/file names of Mac floppy drives (sony.cpp)
	{"cdrom", TYPE_STRING, true},		// Device/file names of Mac CD-ROM drives (cdrom.cpp)
	{"extfs", TYPE_STRING, false},		// Root path of ExtFS (extfs.cpp)
	{"scsi0", TYPE_STRING, false},		// SCSI targets for Mac SCSI ID 0..6 (scsi_*.cpp)
	{"scsi1", TYPE_STRING, false},
	{"scsi2", TYPE_STRING, false},
	{"scsi3", TYPE_STRING, false},
	{"scsi4", TYPE_STRING, false},
	{"scsi5", TYPE_STRING, false},
	{"scsi6", TYPE_STRING, false},
	{"screen", TYPE_STRING, false},		// Video mode (video.cpp)
	{"seriala", TYPE_STRING, false},	// Device name of Mac serial port A (serial_*.cpp)
	{"serialb", TYPE_STRING, false},	// Device name of Mac serial port B (serial_*.cpp)
	{"ether", TYPE_STRING, false},		// Device name of Mac ethernet adapter (ether_*.cpp)
	{"rom", TYPE_STRING, false},		// Path of ROM file (main_*.cpp)
	{"bootdrive", TYPE_INT16, false},	// Boot drive number (main.cpp)
	{"bootdriver", TYPE_INT16, false},	// Boot driver number (main.cpp)
	{"ramsize", TYPE_INT32, false},		// Size of Mac RAM in bytes (main_*.cpp)
	{"frameskip", TYPE_INT32, false},	// Number of frames to skip in refreshed video modes (video_*.cpp)
	{"modelid", TYPE_INT32, false},		// Mac Model ID (Gestalt Model ID minus 6) (rom_patches.cpp)
	{"cpu", TYPE_INT32, false},			// CPU type (0 = 68000, 1 = 68010 etc.) (main.cpp)
	{"fpu", TYPE_BOOLEAN, false},		// Enable FPU emulation (main.cpp)
	{"nocdrom", TYPE_BOOLEAN, false},	// Don't install CD-ROM driver (cdrom.cpp/rom_patches.cpp)
	{"nosound", TYPE_BOOLEAN, false},	// Don't enable sound output (audio_*.cpp)
	{"nogui", TYPE_BOOLEAN, false},		// Disable GUI (main_*.cpp)
	{"idlewait",TYPE_BOOLEAN,true},		// enable idle..
	{"yearoffset",TYPE_INT16,false},	// remove x billion seconds from the clock
	{NULL, TYPE_END, false}	// End of list
};


// Prefs item are stored in a linked list of these nodes
struct prefs_node {
	prefs_node *next;
	const char *name;
	prefs_type type;
	void *data;
};

// List of prefs nodes
static prefs_node *the_prefs;


/*
 *  Initialize preferences
 */

void PrefsInit(void)
{
	// Start with empty list
	the_prefs = NULL;

	// Set defaults
	SysAddSerialPrefs();
	PrefsAddInt16("bootdriver", -33);
	PrefsAddInt16("bootdrive", 0);
	PrefsAddInt32("ramsize", 64 * 1024 * 1024);
	PrefsAddInt32("frameskip", 2);
	PrefsAddInt32("modelid", 29);	// Quadra 800
	PrefsAddInt32("cpu", 4);		// 68040
	PrefsAddBool("fpu", false);		// 68040LC
	PrefsAddBool("nocdrom", false);
	PrefsAddBool("nosound", false);
	PrefsAddBool("nogui", false);
	PrefsAddString("screen","win/1152/870");	//fantastic monitor for the era
	PrefsAddString("rom","Quadra800.rom");
	PrefsAddString("ether","slirp");
	PrefsAddInt16("yearoffset",0);
	AddPlatformPrefsDefaults();

	// Load preferences from settings file
	LoadPrefs();
}


/*
 *  Deinitialize preferences
 */

void PrefsExit(void)
{
	// Free prefs list
	prefs_node *p = the_prefs, *next;
	while (p) {
		next = p->next;
		free((void *)p->name);
		free(p->data);
		delete p;
		p = next;
	}
}


/*
 *  Find preferences descriptor by keyword
 */

static const prefs_desc *find_prefs_desc(const char *name, const prefs_desc *list)
{
	while (list->type != TYPE_ANY) {
		if (strcmp(list->name, name) == 0)
			return list;
		list++;
	}
	return NULL;
}


/*
 *  Set prefs items
 */

static void add_data(const char *name, prefs_type type, void *data, int size)
{
	void *d = malloc(size);
	if (d == NULL)
		return;
	memcpy(d, data, size);
	prefs_node *p = new prefs_node;
	p->next = 0;
	p->name = strdup(name);
	p->type = type;
	p->data = d;
	if (the_prefs) {
		prefs_node *prev = the_prefs;
		while (prev->next)
			prev = prev->next;
		prev->next = p;
	} else
		the_prefs = p;
}

void PrefsAddString(const char *name, const char *s)
{
	add_data(name, TYPE_STRING, (void *)s, strlen(s) + 1);
}

void PrefsAddBool(const char *name, bool b)
{
	add_data(name, TYPE_BOOLEAN, &b, sizeof(bool));
}

void PrefsAddInt16(const char *name, int16 val)
{
	add_data(name, TYPE_INT16, &val, sizeof(int16));
}

void PrefsAddInt32(const char *name, int32 val)
{
	add_data(name, TYPE_INT32, &val, sizeof(int32));
}


/*
 *  Replace prefs items
 */

static prefs_node *find_node(const char *name, prefs_type type, int index = 0)
{
	prefs_node *p = the_prefs;
	int i = 0;
	while (p) {
		if ((type == TYPE_ANY || p->type == type) && !strcmp(p->name, name)) {
			if (i == index)
				return p;
			else
				i++;
		}
		p = p->next;
	}
	return NULL;
}

void PrefsReplaceString(const char *name, const char *s, int index)
{
	prefs_node *p = find_node(name, TYPE_STRING, index);
	if (p) {
		free(p->data);
		p->data = strdup(s);
	} else
		add_data(name, TYPE_STRING, (void *)s, strlen(s) + 1);
}

void PrefsReplaceBool(const char *name, bool b)
{
	prefs_node *p = find_node(name, TYPE_BOOLEAN);
	if (p)
		*(bool *)(p->data) = b;
	else
		add_data(name, TYPE_BOOLEAN, &b, sizeof(bool));
}

void PrefsReplaceInt16(const char *name, int16 val)
{
	prefs_node *p = find_node(name, TYPE_INT16);
	if (p)
		*(int16 *)(p->data) = val;
	else
		add_data(name, TYPE_INT16, &val, sizeof(int16));
}

void PrefsReplaceInt32(const char *name, int32 val)
{
	prefs_node *p = find_node(name, TYPE_INT32);
	if (p)
		*(int32 *)(p->data) = val;
	else
		add_data(name, TYPE_INT32, &val, sizeof(int32));
}


/*
 *  Get prefs items
 */

const char *PrefsFindString(const char *name, int index)
{
	prefs_node *p = find_node(name, TYPE_STRING, index);
	if (p)
		return (char *)(p->data);
	else
		return NULL;
}

bool PrefsFindBool(const char *name)
{
	prefs_node *p = find_node(name, TYPE_BOOLEAN, 0);
	if (p)
		return *(bool *)(p->data);
	else
		return false;
}

int16 PrefsFindInt16(const char *name)
{
	prefs_node *p = find_node(name, TYPE_INT16, 0);
	if (p)
		return *(int16 *)(p->data);
	else
		return 0;
}

int32 PrefsFindInt32(const char *name)
{
	prefs_node *p = find_node(name, TYPE_INT32, 0);
	if (p)
		return *(int32 *)(p->data);
	else
		return 0;
}


/*
 *  Remove prefs items
 */

void PrefsRemoveItem(const char *name, int index)
{
	prefs_node *p = find_node(name, TYPE_ANY, index);
	if (p) {
		free((void *)p->name);
		free(p->data);
		prefs_node *q = the_prefs;
		if (q == p) {
			the_prefs = NULL;
			delete p;
			return;
		}
		while (q) {
			if (q->next == p) {
				q->next = p->next;
				delete p;
				return;
			}
			q = q->next;
		}
	}
}


/*
 *  Load prefs from stream (utility function for LoadPrefs() implementation)
 */

void LoadPrefsFromStream(FILE *f)
{
	char line[256];
	while(fgets(line, 255, f)) {
		// Read line
		int len = strlen(line);
		if (len == 0)
			continue;
		line[len-1] = 0;

		// Comments begin with "#" or ";"
		if (line[0] == '#' || line[0] == ';')
			continue;

		// Kill short lines
		if(strlen(line)<1)
			continue;

		// Terminate string after keyword
		char *p = line;
		while (!isspace(*p)) p++;
		*p++ = 0;

		// Skip whitespace until value
		while (isspace(*p)) p++;
		if (*p == 0)
			continue;
		char *keyword = line;
		char *value = p;
		int32 i = atol(value);

		// Look for keyword first in common item list, then in platform specific list
		const prefs_desc *desc = find_prefs_desc(keyword, common_prefs_items);
		if (desc == NULL)
			desc = find_prefs_desc(keyword, platform_prefs_items);
		if (desc == NULL) {
			//printf("WARNING: Unknown preferences keyword '%s'\n", keyword);
			continue;
		}

		// Add item to prefs
		switch (desc->type) {
			case TYPE_STRING:
				if (desc->multiple)
					PrefsAddString(keyword, value);
				else
					PrefsReplaceString(keyword, value);
				break;
			case TYPE_BOOLEAN:
				PrefsReplaceBool(keyword, !strcmp(value, "true"));
				break;
			case TYPE_INT16:
				PrefsReplaceInt16(keyword, i);
				break;
			case TYPE_INT32:
				PrefsReplaceInt32(keyword, i);
				break;
			default:
				break;
		}
	}
}


/*
 *  Save settings to stream (utility function for SavePrefs() implementation)
 */

static void write_prefs(FILE *f, const prefs_desc *list)
{
	while (list->type != TYPE_ANY) {
		switch (list->type) {
			case TYPE_STRING: {
				int index = 0;
				const char *str;
				while ((str = PrefsFindString(list->name, index++)) != NULL)
					fprintf(f, "%s %s\n", list->name, str);
				break;
			}
			case TYPE_BOOLEAN:
				fprintf(f, "%s %s\n", list->name, PrefsFindBool(list->name) ? "true" : "false");
				break;
			case TYPE_INT16:
				fprintf(f, "%s %d\n", list->name, PrefsFindInt16(list->name));
				break;
			case TYPE_INT32:
				fprintf(f, "%s %ld\n", list->name, PrefsFindInt32(list->name));
				break;
			default:
				break;
		}
		list++;
	}
}

void SavePrefsToStream(FILE *f)
{
	write_prefs(f, common_prefs_items);
	write_prefs(f, platform_prefs_items);
}
